### PR TITLE
feat: eip-2255 implementation

### PR DESCRIPTION
**Pull request by yagopv**
# What it solves
solves #318 

# How this PR fixes it
We are adding support for [EIP-2255](https://eips.ethereum.org/EIPS/eip-2255) (Wallet permissions system)

Adding 3 new methods:

- *getAddressBook* (SDK)
- *wallet_getPermissions* (SDK and Provider)
- *wallet_requestPermissions* (SDK and Provider)

We are changing the `send` function to check if the Method has permissions for being executed

### CODE REVIEW

File: packages/safe-apps-sdk/src/communication/index.ts [hunk 1]
```diff
@@ -55,14 +59,42 @@ class PostMessageCommunicator implements Communicator {
...
+  ): Promise<SuccessResponse<R>> => {
     const request = MessageFormatter.makeRequest(method, params);
 
+    if (Array.isArray(requiredPermissions)) {
+      const hasPermissions = await this.checkPermissions(requiredPermissions);
+
+      if (!hasPermissions) {
+        throw new PermissionsError('Permissions rejected', PERMISSIONS_REQUEST_REJECTED);
+      }
+    }
```
Comment (mikhailxyz) [inline packages/safe-apps-sdk/src/communication/index.ts:90]: Why do we need this logic in the SDK? I think all permission-related checking should be done on the interface side

Comment (mikhailxyz) [reply packages/safe-apps-sdk/src/communication/index.ts:90]: The sdk shouldn't tell if we have permissions or not. The interface should, SDK is only responsible for preparing messages, sending them, and receiving responses

File: packages/safe-apps-provider/package.json [hunk 2]
```diff
@@ -25,7 +25,7 @@
   },
   "homepage": "https://github.com/gnosis/safe-apps-sdk#readme",
   "dependencies": {
-    "@gnosis.pm/safe-apps-sdk": "7.5.0",
+    "@gnosis.pm/safe-apps-sdk": "file:.yalc/@gnosis.pm/safe-apps-sdk",
```
Comment (mikhailxyz) [inline packages/safe-apps-provider/package.json:28]: senor could you remove yalc? yarn will use local packages so no yalc is needed

Comment (yagopv) [reply packages/safe-apps-provider/package.json:28]: yes yes, let me review it again because in the daily i commented it was ready but i had to do the last check as i changed some things this morning :D

File: packages/safe-apps-sdk/src/types/messaging.ts [hunk 3]
```diff
@@ -15,6 +16,9 @@ export interface MethodToResponse {
   [Methods.getSafeBalances]: SafeBalances[];
   [Methods.signMessage]: SendTransactionsResponse;
   [Methods.getEnvironmentInfo]: EnvironmentInfo;
+  [Methods.requestAddressBook]: any;
```
Comment (mikhailxyz) [inline packages/safe-apps-sdk/src/types/messaging.ts]: why any?

File: packages/safe-apps-sdk/src/safe/index.ts [hunk 4]
```diff
@@ -144,6 +146,16 @@ class Safe {
 
     return response.data;
   }
+
+  @requirePermission()
```
[the remark this answers is not in the document]
Comment (yagopv) [reply packages/safe-apps-sdk/src/safe/index.ts:150]: Is missing because we talked about not complicating things in the first iteration and make a 1:1 system, so I'm  only checking if the user has permissions to call the method. So the permission required for `requestAddressBook` is `requestAddressBook` and we are using the `propertyKey` of the method  itself instead passing parameters

File: packages/safe-apps-sdk/src/types/permissions.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  public code: number;
+  public data?: unknown;
+
+  constructor(message: string, code: number, data?: unknown) {
+    super(message);
+
+    this.code = code;
+    this.data = data;
+
+    Object.setPrototypeOf(this, PermissionsError.prototype);
```
Comment (mikhailxyz) [inline packages/safe-apps-sdk/src/types/permissions.ts:34]: Why is this line needed?

File: packages/safe-apps-sdk/src/wallet/wallet.test.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    test('Should send a valid message to the interface', () => {
+      sdkInstance.wallet.getPermissions();
+
+      expect(postMessageSpy).toHaveBeenCalledWith(
+        expect.objectContaining({ method: Methods.wallet_getPermissions, params: undefined }),
+        '*',
+      );
+    });
+
+    test('should resolve the correct getPermissions types', async () => {
```
Comment (mikhailxyz) [inline packages/safe-apps-sdk/src/wallet/wallet.test.ts:28]: This test is obsolete. It checks if the value equals mocked object, it only tests jest's mocking capabilities

File: packages/safe-apps-sdk/src/wallet/wallet.test.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    test('Should send a valid message to the interface', () => {
+      sdkInstance.wallet.requestPermissions([{ requestAddressBook: {} }]);
+
+      expect(postMessageSpy).toHaveBeenCalledWith(
+        expect.objectContaining({ method: Methods.wallet_requestPermissions, params: [{ requestAddressBook: {} }] }),
+        '*',
+      );
+    });
+
+    test('should resolve the correct requestPermissions types', async () => {
```
Comment (mikhailxyz) [inline packages/safe-apps-sdk/src/wallet/wallet.test.ts:64]: same thing with the test here

File: packages/safe-apps-sdk/src/wallet/wallet.test.ts (hunk 6 above)
Comment (mikhailxyz) [reply packages/safe-apps-sdk/src/wallet/wallet.test.ts:28]: If you want to test how it processes the response (this is an excellent idea btw), you can send the response via `window.postMessage`, so it goes through the whole message flow (received by the communicator and then passed back to the calling method)

Comment (mikhailxyz) [reply packages/safe-apps-sdk/src/wallet/wallet.test.ts:28]: This is the entypoint for the message: https://github.com/safe-global/safe-apps-sdk/blob/master/packages/safe-apps-sdk/src/communication/index.ts#L19

File: packages/safe-apps-sdk/src/types/permissions.ts (hunk 5 above)
Comment (yagopv) [reply packages/safe-apps-sdk/src/types/permissions.ts:30]: That's because this https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work

File: packages/safe-apps-sdk/src/wallet/wallet.test.ts (hunk 6 above)
Comment (yagopv) [reply packages/safe-apps-sdk/src/wallet/wallet.test.ts]: Totally right 🙈

Changed

File: packages/safe-apps-sdk/src/types/permissions.ts (hunk 5 above)
Comment (mikhailxyz) [reply packages/safe-apps-sdk/src/types/permissions.ts:30]: Could you please include this as a comment to explain this line?

Comment (yagopv) [reply packages/safe-apps-sdk/src/types/permissions.ts:32]: of course !! done

**Review by mikhailxyz — COMMENTED** [review]
The code looks good, could you please also add documentation for the new methods in the readme?

**Comment by yagopv** [comment]
Sure, added!

**Review by mikhailxyz — APPROVED** [review]
Good job 👍
